### PR TITLE
[Select] Fix mismatch with hidden native select

### LINKED ISSUE #1223: [Select] Underlying native `select` has wrong default value
### Current Behavior

When you set defaultValue in Select component, hidden html select has first option as value instead of defaultValue. So, when you submit uncontrolled form with defaultValue form sends first option of the select as value.
This happens too when you select currently selected option, It sets first option as value.

### Expected behavior

Hidden html select should have defaultValue as value if you do not change the selected option, and if you select the currently selected option, it should not change value.

### Reproducible example

[CodeSandbox Template](https://codesandbox.io/s/select-radix-wrong-value-gl0i4w?file=/App.js)

### PR DESCRIPTION
**Pull request by benoitgrelard**
Closes #1223

> **Note**: easier to review without whitespace

### CODE REVIEW

File: packages/react/select/src/Select.tsx [hunk 1]
```diff
@@ -1010,11 +1044,31 @@ const SelectItemText = React.forwardRef<SelectItemTextElement, SelectItemTextPro
...
+
+    const textContent = itemTextNode?.textContent;
+    const nativeOption = React.useMemo(
+      () => (
+        <option key={itemContext.value} value={itemContext.value} disabled={itemContext.disabled}>
+          {textContent}
+        </option>
+      ),
+      [itemContext.disabled, itemContext.value, textContent]
     );
```
Comment (benoitgrelard) [inline packages/react/select/src/Select.tsx:1064]: I am building the option element directly here, this has 2 advantages in my opinion:
- it makes it more clear/explicit what this is used for, seeing the native option collocated here
- it means I don't need to invent a new data structure like `{ value, label: textContent, disabled }` just to pass it up and interpolate it higher up.

File: packages/react/select/src/Select.tsx [hunk 2]
```diff
@@ -306,34 +293,83 @@ SelectIcon.displayName = ICON_NAME;
...
   (props: ScopedProps<SelectContentProps>, forwardedRef) => {
     const context = useSelectContext(CONTENT_NAME, props.__scopeSelect);
     const [fragment, setFragment] = React.useState<DocumentFragment>();
+    // We set this to true by default so that events bubble to forms without JS (SSR)
+    const isFormControl = context.trigger ? Boolean(context.trigger.closest('form')) : true;
+    const [nativeOptionsSet, setNativeOptionsSet] = React.useState(new Set<ReactOption>());
+
+    const nativeSelectKey = Array.from(nativeOptionsSet.values())
+      .map((option) => option.props.value)
+      .join(';');
```
Comment (andy-hook) [inline packages/react/select/src/Select.tsx:318]: We should try to explain why we need to do this, something like this? (probably bad first attempt):

```
In order for the native select value to update correctly, we need to track and re-render the entire element whenever it's options are changed
```

File: packages/react/select/src/Select.tsx [hunk 3]
```diff
@@ -306,34 +293,83 @@ SelectIcon.displayName = ICON_NAME;
...
+            key={nativeSelectKey}
+            aria-hidden
+            tabIndex={-1}
+            name={context.name}
+            autoComplete={context.autoComplete}
+            value={context.value}
+            // enable form autofill
+            onChange={(event) => context.onValueChange(event.target.value)}
+          >
+            {Array.from(nativeOptionsSet.values())}
```
Comment (andy-hook) [inline packages/react/select/src/Select.tsx:350]: Just curious, is it required to explicitly access `values()`? or is `Array.from(nativeOptionsSet)`  equivalent?

File: packages/react/select/src/Select.tsx [hunk 4]
```diff
@@ -1010,11 +1044,31 @@ const SelectItemText = React.forwardRef<SelectItemTextElement, SelectItemTextPro
     const context = useSelectContext(ITEM_TEXT_NAME, __scopeSelect);
     const contentContext = useSelectContentContext(ITEM_TEXT_NAME, __scopeSelect);
     const itemContext = useSelectItemContext(ITEM_TEXT_NAME, __scopeSelect);
-    const ref = React.useRef<SelectItemTextElement | null>(null);
-    const composedRefs = useComposedRefs(forwardedRef, ref, itemContext.onItemTextChange, (node) =>
-      contentContext.itemTextRefCallback?.(node, itemContext.value, itemContext.disabled)
+    const itemsContext = useSelectNativeOptionsContext(ITEM_TEXT_NAME, __scopeSelect);
```
Comment (andy-hook) [inline packages/react/select/src/Select.tsx:1047]:
```suggestion
    const nativeOptionsContext = useSelectNativeOptionsContext(ITEM_TEXT_NAME, __scopeSelect);
```

Easier to distinguish perhaps?

File: ssr-testing/pages/select.tsx [hunk 5]
```diff
@@ -29,30 +29,32 @@ export default function SelectPage() {
...
-            </Select.Item>
-            <Select.Item value="3">
-              <Select.ItemText>Item 3</Select.ItemText>
-              <Select.ItemIndicator>✔</Select.ItemIndicator>
-            </Select.Item>
-          </Select.Viewport>
-          <Select.ScrollDownButton>▼</Select.ScrollDownButton>
-        </Select.Content>
-      </Select.Root>
+      <form>
```
Comment (andy-hook) [inline ssr-testing/pages/select.tsx:32]: Do you think there's value in rendering both cases for quick manual testing?

File: packages/react/select/src/Select.tsx [hunk 6]
```diff
@@ -1010,11 +1044,31 @@ const SelectItemText = React.forwardRef<SelectItemTextElement, SelectItemTextPro
...
+    const [itemTextNode, setItemTextNode] = React.useState<SelectItemTextElement | null>(null);
+    const composedRefs = useComposedRefs(
+      forwardedRef,
+      (node) => setItemTextNode(node),
+      itemContext.onItemTextChange,
+      (node) => contentContext.itemTextRefCallback?.(node, itemContext.value, itemContext.disabled)
+    );
+
+    const textContent = itemTextNode?.textContent;
```
Comment (andy-hook) [inline packages/react/select/src/Select.tsx:1056]: 👍

File: packages/react/select/src/Select.tsx (hunk 1 above)
Comment (andy-hook) [reply packages/react/select/src/Select.tsx:1064]: haha yeah.. my initial gut reaction was to ask why we don't render closer, I think I'm just conditioned by typically seeing these built closer to the render. But:

> it means I don't need to invent a new data structure like { value, label: textContent, disabled } just to pass it up and interpolate it higher up.

Actually makes much more sense, it's easier to follow when you realise the alternative is just going around the houses for no great reason.

**Review by andy-hook — COMMENTED** [review]
I like this, seems clear enough to me given our shared understanding of the problem. The comments are mostly nits, I think we should add a bit more context for the future with a comment or two though 🙏

File: packages/react/select/src/Select.tsx (hunk 3 above)
Comment (benoitgrelard) [reply packages/react/select/src/Select.tsx:350]: Oh good point, now that it's a set it's probably equivalent, I'll check.

File: ssr-testing/pages/select.tsx (hunk 5 above)
Comment (andy-hook) [reply ssr-testing/pages/select.tsx:32]: > I just thought we should have one with and one without a form

Ah! that's exactly what I was asking for before I took a second to expand the diff and saw that's what's been done here 😅 

All good!

File: packages/react/select/src/Select.tsx [hunk 7]
```diff
@@ -306,34 +293,88 @@ SelectIcon.displayName = ICON_NAME;
...
+      <SelectNativeOptionsProvider
+        scope={props.__scopeSelect}
+        onNativeOptionAdd={React.useCallback((option) => {
+          setNativeOptionsSet((prev) => new Set(prev).add(option));
+        }, [])}
+        onNativeOptionRemove={React.useCallback((option) => {
+          setNativeOptionsSet((prev) => {
+            const itemsSet = new Set(prev);
+            itemsSet.delete(option);
+            return itemsSet;
```
Comment (andy-hook) [inline packages/react/select/src/Select.tsx:340]:
```suggestion
            const optionsSet = new Set(prev);
            optionsSet.delete(option);
            return optionsSet;
```

**Review by andy-hook — APPROVED** [review]

**Pull request merged**